lib/Wikia/src/Service/User/Permissions/PermissionsServiceImpl.php

```diff
@@ -0,0 +1,832 @@
+
+			$fname = __METHOD__;
+			$globalGroups = \WikiaDataAccess::cache(
+				self::getMemcKey( $userId ),
+				\WikiaResponse::CACHE_LONG,
+				function() use ( $userId, $fname ) {
+					$dbr = self::getSharedDB( DB_MASTER );
+					return $dbr->selectFieldValues(
+						'user_groups',
+						'ug_group',
+						[ 'ug_user' => $userId ],
+						$fname
+					);
+				}
+			);
+
+			$globalGroups = array_intersect( $globalGroups, $this->globalGroups );
+			$this->setGlobalUserGroupsArray( $userId, $globalGroups );
+		}
+	}
+
+	/**
+	 * Returns an array of the groups that a particular group can add/remove.
+	 *
+	 * @param $group String: the group to check for whether it can add/remove
+	 * @return Array array( 'add' => array( addablegroups ),
+	 *     'remove' => array( removablegroups ),
+	 *     'add-self' => array( addablegroups to self),
+	 *     'remove-self' => array( removable groups from self) )
+	 */
+	public function getGroupsChangeableByGroup( $group ) {
+		$groups = array( 'add' => array(), 'remove' => array(), 'add-self' => array(), 'remove-self' => array() );
+
+		if ( array_key_exists( $group, $this->groupsAddableByGroup ) ) {
+			$groups['add'] = $this->groupsAddableByGroup[$group];
+		}
+
+		if ( array_key_exists( $group, $this->groupsRemovableByGroup ) ) {
+			$groups['remove'] = $this->groupsRemovableByGroup[$group];
+		}
+
+		if ( array_key_exists( $group, $this->groupsSelfAddableByGroup ) ) {
+			$groups['add-self'] = $this->groupsSelfAddableByGroup[$group];
+		}
+
+		if ( array_key_exists( $group, $this->groupsSelfRemovableByGroup ) ) {
+			$groups['remove-self'] = $this->groupsSelfRemovableByGroup[$group];
+		}
+
+		return $groups;
+	}
+
+	private function isCentralWiki() {
+		global $wgWikiaIsCentralWiki;
+		return (bool)$wgWikiaIsCentralWiki;
+	}
+
+	private function addUserToGlobalGroup( \User $user, $group ) {
+		if ( !$this->isCentralWiki() ) {
+			return false;
+		}
+
+		// Purge cache first so in case of DB failure we don't leave inconsistent data in cache
+		\WikiaDataAccess::cachePurge( self::getMemcKey( $user->getId() ) );
+		$this->invalidateUserGroupsAndPermissions( $user->getId() );
+
+		$dbw = self::getSharedDB( DB_MASTER );
+		if ( $user->getId() ) {
+			$dbw->insert( 'user_groups',
+				[
+					'ug_user'  => $user->getID(),
+					'ug_group' => $group,
+				],
+				__METHOD__
+			);
+		}
+
+		wfRunHooks( 'AfterUserAddGlobalGroup', [ $user, $group ] );
+		return true;
+	}
+
+	private function addUserToLocalGroup( \User $user, $group ) {
+		$dbw = wfGetDB( DB_MASTER );
+		if( $user->getId() ) {
+			$dbw->insert( 'user_groups',
+				array(
+					'ug_user'  => $user->getId(),
+					'ug_group' => $group,
+				),
+				__METHOD__,
+				array( 'IGNORE' ) );
+		}
+		$this->invalidateUserGroupsAndPermissions( $user->getId() );
+
+		return true;
+	}
+
+	private function canGroupBeAdded( \User $userPerformingChange, \User $userToChange, $group ) {
```

Comment:
`$userPerformingChange` is a bit long, how about `$performer`, `$doer` or something else which is shorter?